protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+  be implemented in a manner compatible with the solution outlined here.
+- **Auditability:** An upgrade path should be easy to understand, either by reviewing the source
+  code as in the case of a forthcoming upgrade, or by reviewing the on-chain history along with
+  verified source code.
+- **Atomicity:** It should be possible to complete an upgrade of the entire superchain's L1
+  contracts within one block.
+- **Declarative design:** As much as possible, a developer should be able to simply indicate what
+  bytecode should be deployed, and what storage values (if any) should be initialized or modified.
+
+## Non-goals
+
+This document is not concerned with:
+
+1. upgrading predeploys or other L2 contracts.
+1. modifications to the ownership system, ie, it does not attempt to modify Safes, modules or
+   guards.
+
+More over, it does not present a plan for managing contracts which are not proxied, and not
+Multichain Prepared (MCP-L1) contracts (ie. FaultDisputeGame, AnchorStateRegistry). This
+work will either include, or else depend on, all L1 contracts becoming either:
+
+1. Proxied and MCP ready
+2. Superchain-shared contracts (proxied or not)
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+The proposed solution will perform upgrades by transferring upgrade authorization of all L1
+contracts to the release specific OPCM. Upon completion of the upgrade, this authorization will
+be returned to the Security Council.
+
+Rather than the curent `StorageSetter` based approach which simply deletes the `_initialized` slot,
+allowing for `initialize()` to be called, we take this two-step upgrade further, with a short-lived
+`Populator` contract being temporarily set as the implementation in order to set or modify any
+values as required. This has two significant benefits. First it removes the need to keep a one-time
+`initialize()` function in the run-time logic. Secondly it enables a clean separation of concerns
+between contracts which are newly deployed (requiring all values to be set), and contracts being
+upgraded (requiring only new values to be set).
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
```

Comment:
We should have an onchain way to enforce sequencing between OPCMs when upgrading multiple versions. For example if a chain is on 2.0.0, and 2.1.0 and 2.2.0 have since been released, we want to prevent a footgun where the chain uses OPCM 2.2.0 to upgrade straight to the latest. They should be forced to apply the 2.1.0 and 2.2.0 upgrades (nice to have would be to do both upgrades atomically, but separate transactions seems ok if it's simpler)

Reply:
Good callout. 

I think given we want to use the superchain-registry as the source of truth for OPCM addresses, the approach that comes to my mind is: 

1. Add an `l1ContractsVersion()` getter to the `SystemConfig`. The same getter being added to the OPCM by #158.
2. Add a `previousL1ContractsVersion` value to the OPCM. Have the OPCM check that the current `SystemConfig.l1ContractsVersion() == OPCM.previousL1ContractsVersion` before performing the upgrade.

Reply:
I think we should have an unstructured storage slot that is set in every contract and read that and assert on its value when performing the migration rather than having a new getter on the system config

Reply:
I agree with @blmalone that we should not reverse the previous decision to enforce upgrade ordering off-chain only.

Keeping this open specifically to discuss @tynes comment further.